src/objects/token/admin.py

```diff
@@ -1,11 +1,105 @@
 from django.contrib import admin
+from django.contrib.admin.utils import unquote
 
+from objects.api.serializers import ObjectSerializer
+from objects.core.models import ObjectType
+from objects.utils.admin import EditInlineAdminMixin
+from objects.utils.serializers import build_spec, get_field_names
+
+from .constants import PermissionModes
 from .models import Permission, TokenAuth
 
 
-class PermissionInline(admin.TabularInline):
+@admin.register(Permission)
+class PermissionAdmin(admin.ModelAdmin):
+    list_display = ("token_auth", "object_type", "mode", "use_fields")
+
+    def get_object_fields(self):
+        object_serializer = ObjectSerializer()
+        object_fields = build_spec(get_field_names(object_serializer.fields), sep="__")
+        return object_fields
+
+    def get_data_field_choices(self):
+        data_fields = {}
+        for object_type in ObjectType.objects.all():
+            client = object_type.service.build_client()
+            url = f"{object_type.url}/versions"
+            response = client.request(url, "objectversion_list")
+            if isinstance(response, dict):
+                response = response["results"]
+
+            # to select fields use the latest version
+            # TODO should we include objecttype versions in field-based auth?
+            schema = response[-1]["jsonSchema"]
+            # use only first level of properties
+            properties = list(schema["properties"].keys())
+            data_fields[object_type.id] = {
+                prop: f"record__data__{prop}" for prop in properties
+            }
+        return data_fields
+
+    def get_form_data(self, request, object_id) -> dict:
+        obj = self.get_object(request, unquote(object_id)) if object_id else None
+        ModelForm = self.get_form(request, obj, change=not obj)
+        if request.method == "POST":
+            form = ModelForm(request.POST, request.FILES, instance=obj)
+        else:
+            form = ModelForm(instance=obj)
+        form.is_valid()
+
+        values = {field.name: field.value() for field in form}
+        errors = (
+            {
+                field: [
+                    {"msg": next(iter(error)), "code": error.code} for error in _errors
+                ]
+                for field, _errors in form.errors.as_data().items()
+            }
+            if form.is_bound
+            else {}
+        )
+        return {"values": values, "errors": errors}
+
+    def get_extra_context(self, request, object_id):
+        mode_choices = [("", "---------")] + list(PermissionModes.choices)
+        token_auth_choices = [("", "---------")] + [
+            (token.pk, str(token)) for token in TokenAuth.objects.all()
+        ]
+        object_type_choices = [("", "---------")] + [
```

Comment:
```suggestion
EMPTY_FIELD_CHOICE = ("", "---------")

@admin.register(Permission)
class PermissionAdmin(admin.ModelAdmin):
    list_display = ("token_auth", "object_type", "mode", "use_fields")

    def get_object_fields(self):
        object_serializer = ObjectSerializer()
        object_fields = build_spec(get_field_names(object_serializer.fields), sep="__")
        return object_fields

    def get_data_field_choices(self):
        data_fields = {}
        for object_type in ObjectType.objects.all():
            client = object_type.service.build_client()
            url = f"{object_type.url}/versions"
            response = client.request(url, "objectversion_list")
            if isinstance(response, dict):
                response = response["results"]

            # to select fields use the latest version
            # TODO should we include objecttype versions in field-based auth?
            schema = response[-1]["jsonSchema"]
            # use only first level of properties
            properties = list(schema["properties"].keys())
            data_fields[object_type.id] = {
                prop: f"record__data__{prop}" for prop in properties
            }
        return data_fields

    def get_form_data(self, request, object_id) -> dict:
        obj = self.get_object(request, unquote(object_id)) if object_id else None
        ModelForm = self.get_form(request, obj, change=not obj)
        if request.method == "POST":
            form = ModelForm(request.POST, request.FILES, instance=obj)
        else:
            form = ModelForm(instance=obj)
        form.is_valid()

        values = {field.name: field.value() for field in form}
        errors = (
            {
                field: [
                    {"msg": next(iter(error)), "code": error.code} for error in _errors
                ]
                for field, _errors in form.errors.as_data().items()
            }
            if form.is_bound
            else {}
        )
        return {"values": values, "errors": errors}

    def get_extra_context(self, request, object_id):
        mode_choices = [EMPTY_FIELD_CHOICE] + list(PermissionModes.choices)
        token_auth_choices = [EMPTY_FIELD_CHOICE] + [
            (token.pk, str(token)) for token in TokenAuth.objects.all()
        ]
        object_type_choices = [EMPTY_FIELD_CHOICE] + [
```

Make the dashed string a constant variable to prevent duplication. Could also be inside the class.

Reply:
`EMPTY_FIELD_CHOICE` is added